// gd-sprest/build/mapper/custom/webTemplateExtensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webtemplateextensions = void 0;
var utils_1 = require("../../utils");
/**
 * Web Template Extensions
 * https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-design-rest-api
 */
exports.webtemplateextensions = {
    applySiteDesign: {
        argNames: ["siteDesignId", "webUrl"],
        appendEndpointFl: true,
        name: "ApplySiteDesign",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    addSiteDesignTaskToCurrentWeb: {
        argNames: ["siteDesignId"],
        appendEndpointFl: true,
        name: "AddSiteDesignTaskToCurrentWeb",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    createSiteDesign: {
        argNames: ["info"],
        appendEndpointFl: true,
        name: "CreateSiteDesign",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    createSiteScript: {
        argNames: ["title", "content"],
        appendEndpointFl: true,
        name: "CreateSiteScript(@title)?@title='[[title]]",
        requestType: utils_1.RequestType.PostReplaceWithData
    },
    deleteSiteDesign: {
        argNames: ["id"],
        appendEndpointFl: true,
        name: "DeleteSiteDesign",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    deleteSiteScript: {
        argNames: ["id"],
        appendEndpointFl: true,
        name: "DeleteSiteScript",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    getSiteDesigns: {
        argNames: [],
        appendEndpointFl: true,
        name: "GetSiteDesigns",
        requestType: utils_1.RequestType.Post
    },
    getSiteDesignMetadata: {
        argNames: ["id"],
        appendEndpointFl: true,
        name: "GetSiteDesignMetadata",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    getSiteScripts: {
        argNames: [],
        appendEndpointFl: true,
        name: "GetSiteScripts",
        requestType: utils_1.RequestType.Post
    },
    getSiteScriptFromWeb: {
        argNames: ["webUrl", "info"],
        appendEndpointFl: true,
        name: "GetSiteScriptFromWeb",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    getSiteScriptFromList: {
        argNames: ["listUrl"],
        appendEndpointFl: true,
        name: "GetSiteScriptFromList",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    getSiteScriptMetadata: {
        argNames: ["id"],
        appendEndpointFl: true,
        name: "GetSiteScriptMetadata",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    getSiteDesignRights: {
        argNames: ["id"],
        appendEndpointFl: true,
        name: "GetSiteDesignRights",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    grantSiteDesignRights: {
        argNames: ["id", "principalNames", "grantedRights"],
        appendEndpointFl: true,
        name: "GrantSiteDesignRights",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    revokeSiteDesignRights: {
        argNames: ["id", "principalNames"],
        appendEndpointFl: true,
        name: "RevokeSiteDesignRights",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    updateSiteDesign: {
        argNames: ["updateInfo"],
        appendEndpointFl: true,
        name: "UpdateSiteDesign",
        requestType: utils_1.RequestType.PostWithArgsInBody
    },
    updateSiteScript: {
        argNames: ["updateInfo"],
        appendEndpointFl: true,
        name: "UpdateSiteScript",
        requestType: utils_1.RequestType.PostWithArgsInBody
    }
};
