// gd-bs/build/components/form/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormValidationTypes = exports.FormControlTypes = void 0;
/**
 * Form Control Types
 */
var FormControlTypes;
(function (FormControlTypes) {
    FormControlTypes[FormControlTypes["Checkbox"] = 1] = "Checkbox";
    FormControlTypes[FormControlTypes["ColorPicker"] = 2] = "ColorPicker";
    FormControlTypes[FormControlTypes["Email"] = 3] = "Email";
    FormControlTypes[FormControlTypes["Datalist"] = 4] = "Datalist";
    FormControlTypes[FormControlTypes["Dropdown"] = 5] = "Dropdown";
    FormControlTypes[FormControlTypes["File"] = 6] = "File";
    FormControlTypes[FormControlTypes["ListBox"] = 7] = "ListBox";
    FormControlTypes[FormControlTypes["MultiCheckbox"] = 8] = "MultiCheckbox";
    FormControlTypes[FormControlTypes["MultiDropdown"] = 9] = "MultiDropdown";
    FormControlTypes[FormControlTypes["MultiListBox"] = 10] = "MultiListBox";
    FormControlTypes[FormControlTypes["MultiRadio"] = 11] = "MultiRadio";
    FormControlTypes[FormControlTypes["MultiSwitch"] = 12] = "MultiSwitch";
    FormControlTypes[FormControlTypes["Password"] = 13] = "Password";
    FormControlTypes[FormControlTypes["Radio"] = 14] = "Radio";
    FormControlTypes[FormControlTypes["Range"] = 15] = "Range";
    FormControlTypes[FormControlTypes["Readonly"] = 16] = "Readonly";
    FormControlTypes[FormControlTypes["Switch"] = 17] = "Switch";
    FormControlTypes[FormControlTypes["TextArea"] = 18] = "TextArea";
    FormControlTypes[FormControlTypes["TextField"] = 19] = "TextField";
})(FormControlTypes = exports.FormControlTypes || (exports.FormControlTypes = {}));
/**
 * Form Validation Types
 */
var FormValidationTypes;
(function (FormValidationTypes) {
    FormValidationTypes[FormValidationTypes["Default"] = 1] = "Default";
    FormValidationTypes[FormValidationTypes["Tooltip"] = 2] = "Tooltip";
})(FormValidationTypes = exports.FormValidationTypes || (exports.FormValidationTypes = {}));
