// dattatable/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var Common = require("./common");
__exportStar(require("./common"), exports);
var dashboard_1 = require("./dashboard");
__exportStar(require("./dashboard"), exports);
var documents_1 = require("./documents");
__exportStar(require("./documents"), exports);
var itemForm_1 = require("./itemForm");
__exportStar(require("./itemForm"), exports);
/** Styling */
require("./styles");
/** Global Variable */
window["DattaTable"] = {
    ActionButtonTypes: documents_1.ActionButtonTypes,
    Common: Common,
    Dashboard: dashboard_1.Dashboard,
    Documents: documents_1.Documents,
    ItemForm: itemForm_1.ItemForm
};
