// dattatable/build/common/modal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Modal = void 0;
var gd_sprest_bs_1 = require("gd-sprest-bs");
/**
 * Modal
 */
var Modal = /** @class */ (function () {
    // Constructor
    function Modal() {
        // Render the canvas
        Modal.render();
    }
    Object.defineProperty(Modal, "BodyElement", {
        get: function () { return this._elBody; },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Modal, "FooterElement", {
        get: function () { return this._elFooter; },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Modal, "HeaderElement", {
        get: function () { return this._elHeader; },
        enumerable: false,
        configurable: true
    });
    // Clears the canvas form
    Modal.clear = function () {
        // Clear the header, body and footer
        this.setHeader("");
        this.setBody("");
        this.setFooter("");
    };
    // Hides the modal
    Modal.hide = function () { this._modal.hide(); };
    // Renders the canvas
    Modal.render = function () {
        var _this = this;
        // Create the element
        var el = document.createElement("div");
        el.id = "core-modal";
        // Ensure the body exists
        if (document.body) {
            // Append the element
            document.body.appendChild(el);
        }
        else {
            // Create an event
            window.addEventListener("load", function () {
                // Append the element
                document.body.appendChild(el);
            });
        }
        // Render the canvas
        this._modal = gd_sprest_bs_1.Components.Modal({
            el: el,
            isCentered: true,
            type: gd_sprest_bs_1.Components.ModalTypes.Large,
            options: {
                autoClose: false,
                backdrop: true,
                keyboard: true
            },
            onRenderBody: function (el) { _this._elBody = el; },
            onRenderFooter: function (el) { _this._elFooter = el; },
            onRenderHeader: function (el) { _this._elHeader = el.querySelector(".modal-title"); ; },
            onClose: function () {
                // Call the close event
                _this._onCloseEvent ? _this._onCloseEvent() : null;
            }
        });
    };
    // Sets the auto close flag
    Modal.setAutoClose = function (value) { this._modal.setAutoClose(value); };
    // Sets the body
    Modal.setBody = function (content) {
        // Clear the body
        while (this._elBody.firstChild) {
            this._elBody.removeChild(this._elBody.firstChild);
        }
        // See if content exists
        if (content) {
            // See if this is text
            if (typeof (content) == "string") {
                // Set the html
                this._elBody.innerHTML = content;
            }
            else {
                // Append the element
                this._elBody.appendChild(content);
            }
        }
    };
    // Sets the close event
    Modal.setCloseEvent = function (event) { this._onCloseEvent = event; };
    // Sets the footer
    Modal.setFooter = function (content) {
        // Clear the body
        while (this._elFooter.firstChild) {
            this._elFooter.removeChild(this._elFooter.firstChild);
        }
        // See if content exists
        if (content) {
            // See if this is text
            if (typeof (content) == "string") {
                // Set the html
                this._elFooter.innerHTML = content;
            }
            else {
                // Append the element
                this._elFooter.appendChild(content);
            }
        }
    };
    // Sets the header
    Modal.setHeader = function (content) {
        // Clear the body
        while (this._elHeader.firstChild) {
            this._elHeader.removeChild(this._elHeader.firstChild);
        }
        // See if content exists
        if (content) {
            // See if this is text
            if (typeof (content) == "string") {
                // Set the html
                this._elHeader.innerHTML = content;
            }
            else {
                // Append the element
                this._elHeader.appendChild(content);
            }
        }
    };
    // Sets the modal type
    Modal.setType = function (type) { this._modal.setType(type); };
    // Shows the modal
    Modal.show = function () { this._modal.show(); };
    Modal._modal = null;
    Modal._onCloseEvent = null;
    // Modal Body
    Modal._elBody = null;
    // Modal Footer
    Modal._elFooter = null;
    // Modal Header
    Modal._elHeader = null;
    return Modal;
}());
exports.Modal = Modal;
// Create an instance of the modal
new Modal();
